File: day2_client/sse.py
class Parser:
    def __init__(self) -> None:
        self._buffer = bytearray()
        self._index = 0
        self._line_buffer = b""
        self._event_lines: list[bytes] = []
        self._pending_cr = False
        self._emitted_payloads: list[str] = []

    def feed(self, chunk: bytes) -> list[str]:
        if chunk:
            self._buffer.extend(chunk)
        self._process_buffer()
        return self._drain_emitted_payloads()

    def eof(self) -> list[str]:
        if self._pending_cr:
            self._pending_cr = False
            self._finish_line()
        if self._line_buffer:
            self._finish_line()
        if self._event_lines:
            self._emit_current_event()
        return self._drain_emitted_payloads()

    def _process_buffer(self) -> None:
        while self._index < len(self._buffer):
            if self._pending_cr:
                if self._buffer[self._index : self._index + 1] == b"\n":
                    self._index += 1
                    self._finish_line()
                    self._pending_cr = False
                    continue

                self._pending_cr = False
                self._finish_line()
                continue

            byte = bytes(self._buffer[self._index : self._index + 1])
            self._index += 1

            if byte == b"\r":
                self._pending_cr = True
            elif byte == b"\n":
                self._finish_line()
            else:
                self._line_buffer += byte

        self._compact_buffer()

    def _compact_buffer(self) -> None:
        if self._index > 0:
            del self._buffer[: self._index]
            self._index = 0

    def _finish_line(self) -> None:
        if self._line_buffer:
            self._event_lines.append(self._line_buffer)
            self._line_buffer = b""
            return

        self._emit_current_event()

    def _emit_current_event(self) -> None:
        if self._event_lines:
            payload = self._parse_event(self._event_lines)
            if payload is not None:
                self._emitted_payloads.append(payload)
            self._event_lines = []

    def _parse_event(self, lines: list[bytes]) -> str | None:
        data_lines: list[str] = []
        for line in lines:
            if line.startswith(b":"):
                continue
            if line.startswith(b"data:"):
                payload = line[len(b"data:") :]
                if payload.startswith(b" "):
                    payload = payload[1:]
                data_lines.append(payload.decode("utf-8"))

        if data_lines:
            return "\n".join(data_lines)
        return None

    def _drain_emitted_payloads(self) -> list[str]:
        payloads = self._emitted_payloads
        self._emitted_payloads = []
        return payloads

File: ts/sse.ts
/**
 * The same SSE state machine as `day2_client/sse.py`, line for line, so the two
 * can be compared honestly.
 *
 * One real difference. Python buffers *bytes* and decodes each data line once
 * it is whole. Here the decoder runs first, so it must be told the input is a
 * stream:
 *
 *   new TextDecoder().decode(chunk)                  // corrupts split UTF-8
 *   new TextDecoder().decode(chunk, {stream: true})  // holds the partial bytes
 *
 * A network chunk can end in the middle of a multi-byte character just as
 * easily as in the middle of an event. Without `{stream: true}` you get U+FFFD
 * in the transcript, which — like the `\n\n` bug — will not show up in English
 * test data and will show up the first time a user types an emoji.
 */
export class SseParser {
  #lineBuffer = "";
  #eventLines: string[] = [];
  #pendingCr = false;
  #decoder = new TextDecoder("utf-8");

  feed(chunk: Uint8Array): string[] {
    return this.#consume(this.#decoder.decode(chunk, { stream: true }));
  }

  eof(): string[] {
    const payloads = this.#consume(this.#decoder.decode());
    if (this.#pendingCr) {
      this.#pendingCr = false;
      this.#finishLine(payloads);
    }
    if (this.#lineBuffer) this.#finishLine(payloads);
    if (this.#eventLines.length) this.#emit(payloads);
    return payloads;
  }

  #consume(text: string): string[] {
    const payloads: string[] = [];

    for (const char of text) {
      // A lone \r is a valid line terminator, so it can only be resolved once
      // the next character arrives — which may be in the next network chunk.
      if (this.#pendingCr) {
        this.#pendingCr = false;
        this.#finishLine(payloads);
        if (char === "\n") continue;
      }

      if (char === "\r") this.#pendingCr = true;
      else if (char === "\n") this.#finishLine(payloads);
      else this.#lineBuffer += char;
    }

    return payloads;
  }

  #finishLine(payloads: string[]): void {
    if (this.#lineBuffer) {
      this.#eventLines.push(this.#lineBuffer);
      this.#lineBuffer = "";
      return;
    }
    // An empty line is the event terminator, not an empty line.
    this.#emit(payloads);
  }

  #emit(payloads: string[]): void {
    if (!this.#eventLines.length) return;

    const data: string[] = [];
    for (const line of this.#eventLines) {
      if (line.startsWith(":")) continue;
      if (!line.startsWith("data:")) continue;
      let value = line.slice("data:".length);
      if (value.startsWith(" ")) value = value.slice(1);
      data.push(value);
    }

    this.#eventLines = [];
    if (data.length) payloads.push(data.join("\n"));
  }
}

/** Pull assistant text out of one OpenAI-compatible chunk payload. */
export function contentOf(payload: string): string | undefined {
  if (payload === "[DONE]") return undefined;
  try {
    const chunk = JSON.parse(payload);
    const content = chunk?.choices?.[0]?.delta?.content;
    return typeof content === "string" && content ? content : undefined;
  } catch {
    return undefined;
  }
}
